### PR TITLE
[4.4] Fixes to Form Field classes from static code analysis

### PR DESCRIPTION
**Pull request by Hackwar**
### Summary of Changes
These are several fixes to the `Joomla\CMS\Form\Field` classes based on static code analysis. 

### Testing Instructions
Codereview.

### Link to documentations
Please select:
- [ ] Documentation link for docs.joomla.org: <link>
- [X] No documentation changes for docs.joomla.org needed

- [ ] Pull Request link for manual.joomla.org: <link>
- [X] No documentation changes for manual.joomla.org needed

### CODE REVIEW

File: libraries/src/Form/Field/LanguageField.php [hunk 1]
```diff
@@ -69,27 +69,27 @@ function ($a, $b) {
...
-                    break;
-                case 'active':
-                case 'auto':
-                    $lang        = Factory::getLanguage();
-                    $this->value = $lang->getTag();
-                    break;
-                default:
-                    break;
-            }
+        $langParams = ComponentHelper::getParams('com_languages');
```
Comment (laoneo) [inline libraries/src/Form/Field/LanguageField.php:72]: Why did you remove here the if?

File: libraries/src/Form/Field/ListField.php [hunk 2]
```diff
@@ -79,31 +79,31 @@ protected function getOptions()
...
-                    continue;
-                }
+            $requires = explode(',', (string) $option['requires']);
 
-                // Requires associations
-                if (\in_array('associations', $requires) && !Associations::isEnabled()) {
-                    continue;
-                }
+            // Requires multilanguage
+            if (\in_array('multilanguage', $requires) && !Multilanguage::isEnabled()) {
```
Comment (laoneo) [inline libraries/src/Form/Field/ListField.php:85]: Are you sure?

File: libraries/src/Form/Field/LanguageField.php (hunk 1 above)
Comment (Hackwar) [reply libraries/src/Form/Field/LanguageField.php:72]: Because that line 72 always evaluates to true. It is simply useless.

File: libraries/src/Form/Field/ListField.php (hunk 2 above)
Comment (Hackwar) [reply libraries/src/Form/Field/ListField.php:85]: This also always evaluates to true. Besides the fact that at some point we had the rule "no assignments in if-clauses".

File: libraries/src/Form/Field/CalendarField.php [hunk 3]
```diff
@@ -243,8 +243,8 @@ public function setup(\SimpleXMLElement $element, $value, $group = null)
             $this->filltable    = (string) $this->element['filltable'] ? (string) $this->element['filltable'] : 'true';
             $this->timeformat   = (int) $this->element['timeformat'] ? (int) $this->element['timeformat'] : 24;
             $this->singleheader = (string) $this->element['singleheader'] ? (string) $this->element['singleheader'] : 'false';
-            $this->minyear      = \strlen((string) $this->element['minyear']) ? (string) $this->element['minyear'] : null;
-            $this->maxyear      = \strlen((string) $this->element['maxyear']) ? (string) $this->element['maxyear'] : null;
+            $this->minyear      = \strlen((string) $this->element['minyear']) ? (int) $this->element['minyear'] : null;
```
Comment (laoneo) [inline libraries/src/Form/Field/CalendarField.php:246]: I'm a bit hesitant to add this change in a minor. I do fully agree that it is correct, but should go into 5.0.

File: libraries/src/Form/Field/ListField.php (hunk 2 above)
Comment (laoneo) [inline libraries/src/Form/Field/ListField.php:85]: Explode can also return false in PHP < 8, read the changelog here https://www.php.net/manual/en/function.explode.php

File: libraries/src/Form/Field/TextField.php [hunk 4]
```diff
@@ -186,16 +186,14 @@ public function setup(\SimpleXMLElement $element, $value, $group = null)
             $inputmode       = preg_replace('/\s+/', ' ', trim($inputmode));
             $inputmode       = explode(' ', $inputmode);
 
-            if (!empty($inputmode)) {
-                $defaultInputmode = \in_array('default', $inputmode) ? Text::_('JLIB_FORM_INPUTMODE') . ' ' : '';
+            $defaultInputmode = \in_array('default', $inputmode) ? Text::_('JLIB_FORM_INPUTMODE') . ' ' : '';
```
Comment (laoneo) [inline libraries/src/Form/Field/TextField.php:189]: See comment above, in some cases explode can return false

File: libraries/src/Form/Field/TimeField.php [hunk 5]
```diff
@@ -113,9 +107,9 @@ public function setup(\SimpleXMLElement $element, $value, $group = null)
 
         if ($return) {
             // It is better not to force any default limits if none is specified
-            $this->max  = isset($this->element['max']) ? (string) $this->element['max'] : null;
-            $this->min  = isset($this->element['min']) ? (string) $this->element['min'] : null;
-            $this->step = isset($this->element['step']) ? (float) $this->element['step'] : null;
+            $this->max  = isset($this->element['max']) ? (int) $this->element['max'] : null;
```
Comment (laoneo) [inline libraries/src/Form/Field/TimeField.php:110]: Should also go into 5

File: libraries/src/Form/Field/ListField.php (hunk 2 above)
Comment (Hackwar) [reply libraries/src/Form/Field/ListField.php:85]: It only returns false when the separator is an empty string. Since we can see in line 82 that the separator is definitely not empty, we can rule that one out.

File: libraries/src/Form/Field/TextField.php (hunk 4 above)
Comment (Hackwar) [reply libraries/src/Form/Field/TextField.php:189]: See my response above. We see in line 187 that the separator is not empty.

Comment (laoneo) [reply libraries/src/Form/Field/TextField.php:189]: ok

File: libraries/src/Form/Field/ListField.php (hunk 2 above)
Comment (laoneo) [reply libraries/src/Form/Field/ListField.php:85]: ok

File: libraries/src/Form/Field/CalendarField.php (hunk 3 above)
Comment (Hackwar) [reply libraries/src/Form/Field/CalendarField.php:246]: The thing is, that using a string here will result in an error, so this is actually fixing a bug.

File: libraries/src/Form/Field/TimeField.php (hunk 5 above)
Comment (Hackwar) [reply libraries/src/Form/Field/TimeField.php:110]: Again, this is also fixing bugs. If we allow strings here, the resulting output would be invalid.

**Comment by brianteeman** [comment]
and this breaks everything :(

**Comment by laoneo** [comment]
Can you post here the error message you get? It looks like there is an issue with your setup. Also make sure you run `composer install` and `npm ci` on a clean branch 4.4-dev. Because the 130 system tests were passing fine, so a lot of views should work .

**Comment by brianteeman** [comment]
The obvious error in the modules is fixed by #40403 but there are a lot more changes in this PR that need real testing not just a review by one person

**Comment by laoneo** [comment]
We are aware of the module error and @Hackwar cam up very quickly with a solution with #40403. But you wrote that everything is broken, so it looks like more an issue with your setup.

It was not only me who reviewed it, some other maintainers did as well. That's why I wanted to merge it early in the 4.4 release lifecycle to spot and fix such issues quickly.
